### PR TITLE
change app.Router interface

### LINKED ISSUE #817: Change router.Handle notation
As a developer, I want to change `router.Handle` notation so that handler registration and message declaration is less verbose.

The current [`router.Handle`](https://github.com/iov-one/weave/blob/master/app/router.go#L42) notation is
```go
func (r Router) Handle(path string, h weave.Handler)
```
First argument is the `path`, that is the same as the result of `Msg.Path()` that registered handler handles.

The above notation is correct and it can be improved. The current notation require us to [create a message in place and call `Path()`](https://github.com/iov-one/weave/blob/master/cmd/bnsd/x/username/handlers.go#L21-L23) to get the path value or to [declare the path value as a `const`](https://github.com/iov-one/weave/blob/master/x/paychan/msg.go#L20-L22). Both approaches are valid and can be improved.
Message path is used only by the [`app.Router`](https://github.com/iov-one/weave/blob/master/app/router.go#L26).

## Acceptance criteria

- [x] change the notation to require `weave.Msg` as the first arugment
```go
func (r Router) Handle(m weave.Msg, h weave.Handler)
```
- [x] adjust all handler registrations
- [x] remove all `const` declaration of the message paths. `Msg.Path` method should be the only place where the path is declared and available

### PR DESCRIPTION
**Pull request by husio**
Change the `app.Router` to require a message reference instead of the
message path as the first argument.

Remove all `const` declaration of message paths as they are no longer
needed and can be embeded directly in message `Path` method.

resolve #817

### CODE REVIEW

File: app/router.go [hunk 1]
```diff
@@ -27,75 +24,66 @@ type Router struct {
...
+func NewRouter() *Router {
+	return &Router{
+		routes: make(map[string]weave.Handler),
 	}
 }
 
-// Handle adds a new Handler for the given path.
-// panics if another Handler was already registered
-func (r Router) Handle(path string, h weave.Handler) {
+// Handle implementes weave.Registry interface.
```
Comment (ethanfrey) [inline app/router.go]: sp: implements

File: app/router.go [hunk 2]
```diff
@@ -27,75 +24,66 @@ type Router struct {
...
+	return notFoundHandler(path)
 }
 
 // Check dispatches to the proper handler based on path
-func (r Router) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
-	msg, _ := tx.GetMsg()
-	if msg == nil {
-		return nil, errors.Wrap(errors.ErrInput, "unable to decode")
+func (r *Router) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	msg, err := tx.GetMsg()
```
Comment (ethanfrey) [inline app/router.go]: Nice unrelated cleanup here

File: app/router.go [hunk 3]
```diff
@@ -27,75 +24,66 @@ type Router struct {
...
-// Handler returns the registered Handler for this path.
-// If no path is found, returns a noSuchPath Handler
-// Always returns a non-nil Handler
-func (r Router) Handler(path string) weave.Handler {
-	h, ok := r.routes[path]
-	if !ok {
-		return noSuchPathHandler{path}
+// handler returns the registered Handler for this path. If no path is found,
+// returns a noSuchPath Handler.  This method always returns a non-nil Handler.
+func (r *Router) handler(m weave.Msg) weave.Handler {
```
Comment (ethanfrey) [inline app/router.go]: Fair enough, I guess it doesn't need to be public

File: cmd/bnsd/x/username/handlers.go [hunk 4]
```diff
@@ -18,9 +18,9 @@ func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
 	r = migration.SchemaMigratingRegistry("username", r)
 
 	b := NewTokenBucket()
-	r.Handle(RegisterTokenMsg{}.Path(), &registerTokenHandler{auth: auth, bucket: b})
-	r.Handle(TransferTokenMsg{}.Path(), &transferTokenHandler{auth: auth, bucket: b})
-	r.Handle(ChangeTokenTargetsMsg{}.Path(), &changeTokenTargetsHandler{auth: auth, bucket: b})
+	r.Handle(&RegisterTokenMsg{}, &registerTokenHandler{auth: auth, bucket: b})
```
Comment (ethanfrey) [inline cmd/bnsd/x/username/handlers.go]: Yeah, nicer.
You could also do `(*RegisterTokenMsg)(nil)` if you want to get really carried away, but I think this is more readable

File: x/aswap/msg.go [hunk 5]
```diff
@@ -1,52 +1,29 @@
...
-	return pathCreateSwap
-}
-
-func (ReleaseMsg) Path() string {
-	return pathReleaseSwap
-}
-
-func (ReturnSwapMsg) Path() string {
-	return pathReturnSwap
+	return "aswap/create"
```
Comment (ethanfrey) [inline x/aswap/msg.go]: :+1: for inlining all this

File: x/aswap/msg.go [hunk 6]
```diff
@@ -1,52 +1,29 @@
...
-	pathReturnSwap  = "aswap/return"
-
 	maxMemoSize int = 128
 	// preimage size in bytes
 	preimageSize int = 32
 	// preimageHash size in bytes
 	preimageHashSize int = 32
 )
 
-var _ weave.Msg = (*CreateMsg)(nil)
```
Comment (ethanfrey) [inline x/aswap/msg.go]: I like these to help enforce the Msg interface here, rather than some random error when type casting in another file. Did you move these "assertions" somewhere else, or just delete them?

I feel when you make a struct with the sole purpose of fulfilling an interface, declaring that in a language the compiler understand is a "good thing"

File: x/currency/msg.go [hunk 7]
```diff
@@ -11,8 +10,6 @@ func init() {
 	migration.MustRegister(1, &CreateMsg{}, migration.NoModification)
 }
 
-var _ weave.Msg = (*CreateMsg)(nil)
```
Comment (ethanfrey) [inline x/currency/msg.go]: Again, why this is removed?

File: app/router.go (hunk 2 above)
Comment (husio) [reply app/router.go]: And a :bug: fix!

File: x/aswap/msg.go (hunk 6 above)
Comment (husio) [reply x/aswap/msg.go]: I was deleting all of them because I think there are enough uses of message and model interface to validate interface implementation. There is migration in the same file and handler and model registrations.
I used to put those everywhere before we had those checks.

Should they stay anyway?

Comment (ruseinov) [reply x/aswap/msg.go]: I quite like interface validation on top, it gives a nice indication of what needs to be implemented and show and error in the IDE if that's not the case. 

Very nice for newcomers too.

**Comment by ethanfrey** [comment]
Sorry about the last comment. When reflecting on it last night, I realized the new router handle method that takes Msg will enforce this... So this is conpiler checked in the package and those lines are no longer needed.

I just wanted to ensure enforcement was there

File: cmd/bnsd/x/username/handlers.go [hunk 8]
```diff
@@ -18,9 +18,9 @@ func RegisterRoutes(r weave.Registry, auth x.Authenticator) {
...
+	r.Handle(&TransferTokenMsg{}, &transferTokenHandler{auth: auth, bucket: b})
```
Comment (ethanfrey) [inline cmd/bnsd/x/username/handlers.go]: And this actually enforces `*TransferTokenMsg` is a `weave.Msg`, eliminating need for my home-grown compiler assertions.

Comment (husio) [reply cmd/bnsd/x/username/handlers.go]: Yes, this is what I was trying to explain :) Because I do not feel strongly about those type assertions, I have put them all back.
